sys/drv/bcm2835_pic.c

```diff
@@ -0,0 +1,204 @@
+#define KL_LOG KL_DEV
+#include <sys/bus.h>
+#include <sys/devclass.h>
+#include <sys/errno.h>
+#include <sys/fdt.h>
+#include <sys/interrupt.h>
+#include <sys/klog.h>
+#include <sys/libkern.h>
+#include <dev/bcm2835reg.h>
+
+/*
+ * located at BCM2835_ARMICU_BASE
+ * accessed by BCM2835_PERIPHERALS_BASE NOT by BCM2835_PERIPHERALS_BASE_BUS
+ * 32 GPU0 interrupts -- 0x204 offset from address; 0 in table
+ * 32 GPU1 interrupts -- 0x208 offset from address; 32 in table
+ * 32 basic interrupts -- 0x200 offset from address; 64 in table
+ */
+
+/* NOTE: the values of the following enum must not be modified! */
+typedef enum bcm2835_intrtype {
+  BCM2835_INTRTYPE_BASIC,
+  BCM2835_INTRTYPE_GPU0,
+  BCM2835_INTRTYPE_GPU1,
+  BCM2835_INTRTYPE_CNT
+} bcm2835_intrtype_t;
+
+#define BCM2835_NIRQPERTYPE (BCM2835_NIRQ / BCM2835_INTRTYPE_CNT)
+
+typedef struct bcm2835_pic_state {
+  rman_t rm;
+  resource_t *mem;
+  resource_t *irq;
+  intr_event_t *intr_event[BCM2835_NIRQ];
+} bcm2835_pic_state_t;
+
+#define in4(addr) bus_read_4(bcm2835_pic->mem, (addr))
+#define out4(addr, val) bus_write_4(bcm2835_pic->mem, (addr), (val))
+
+static void bcm2835_pic_disable_irq(intr_event_t *ie) {
+  bcm2835_pic_state_t *bcm2835_pic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  bus_size_t reg = BCM2835_INTC_IRQBDISABLE;
+  if (irq < BCM2835_INT_GPU1BASE)
+    reg = BCM2835_INTC_IRQ1DISABLE;
+  else if (irq < BCM2835_INT_BASICBASE)
+    reg = BCM2835_INTC_IRQ2DISABLE;
+
+  uint32_t en = in4(reg);
+  out4(reg, en & ~(1 << irq));
+}
+
+static void bcm2835_pic_enable_irq(intr_event_t *ie) {
+  bcm2835_pic_state_t *bcm2835_pic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  bus_size_t reg = BCM2835_INTC_IRQBENABLE;
+  if (irq < BCM2835_INT_GPU1BASE)
+    reg = BCM2835_INTC_IRQ1ENABLE;
+  else if (irq < BCM2835_INT_BASICBASE)
+    reg = BCM2835_INTC_IRQ2ENABLE;
```

Comment:
It's the same code (I mean logic) as before. If you want you can consider adding helper function irq -> reg.

Reply:
I could introduce `intrtype_to_reg(intrtype, regtype)` where `regtype` is an enum (`enable`, `disable`, `pending`). On the other hand, FTTB, I only need to translate `intrtype` once to enable register and once to disable register so maybe that is too generic (and requires a new type definition). @cahirwpz what do you think?

Reply:
Let's not complicate. This code is readable enough for us to track. Let's leave it simple.